Add support for RFC8439 variant of ChaCha20

sipa opened the pull request:
Based on and replaces part of #25361, part of the BIP324 project (#27634). See also #19225 for background.

There are two variants of ChaCha20 in use. The currently implemented one uses a 64-bit nonce and a 64-bit block counter, while the one used in RFC8439 (and thus BIP324) uses a 96-bit nonce and 32-bit block counter. This PR changes the logic to use the 96-bit nonce variant, though in a way that's compatible with >256 GiB output (by automatically incrementing the first 32-bit part of the nonce when the block counter overflows).

For those who reviewed the original PR, the biggest change is here that the 96-bit nonce is passed as a Nonce96 type (pair of 32-bit + 64-bit integer) rather than a 12-byte array.

sipa committed:
crypto: Implement RFC8439-compatible variant of ChaCha20

There are two variants of ChaCha20 in use. The original one uses a 64-bit
nonce and a 64-bit block counter, while the one used in RFC8439 uses a
96-bit nonce and 32-bit block counter. This commit changes the interface
to use the 96/32 split (but automatically incrementing the first 32-bit
part of the nonce when the 32-bit block counter overflows, so to retain
compatibility with >256 GiB output).

Simultaneously, also merge the SetIV and Seek64 functions, as we almost
always call both anyway.

Co-authored-by: dhruv <856960+dhruv@users.noreply.github.com>

Sjors commented:
#19225 :-)

theStack commented on src/test/crypto_tests.cpp:
```diff
@@ -184,6 +184,33 @@ static void TestChaCha20(const std::string &hex_message, const std::string &hexk
+        BOOST_CHECK_EQUAL(HexStr(keystream), hex_expected_keystream);
+    }
+
+    if (!hex_input.empty()) {
+        assert(hex_input.size() == hex_expected_output.size());
```
tiny nit (probably only worth changing if you have to retouch): could check this assert unconditionally, to not allow test cases where `hex_input` is empty but `hex_expected_output` is not

theStack reviewed (commented):
Concept ACK

theStack reviewed (approved):
Code-review ACK 200e1a292b277d6c32dbd6081d38b715cbc079cb

Verified that the implementation matches the RFC8439 specification (https://datatracker.ietf.org/doc/html/rfc8439#section-2.3) and checked that the test vectors match the ones from the linked sources (https://tools.ietf.org/html/draft-agl-tls-chacha20poly1305-04#section-7 and  https://datatracker.ietf.org/doc/html/rfc8439#page-30).

Note for other reviewers (that get as easily confused as me):  we don't store the constants in the `input` array, so our indices are 4 less, e.g. `input[8]` in our code corresponds to `Word 12` in the RFC.

achow101 commented:
Is the current ChaCha20 implementation used for anything that needs backwards compatibility?

achow101 commented:
Assuming it doesn't break a whole lot of tests, I would prefer to change it always use the new version as it removes the need to check that everything is using the right variant.

sipa committed:
tests: improve ChaCha20 unit tests

achow101 commented:
ACK 7f2a985147ef541123c65d5db1c3fc3e533fd4ce

This PR fundamentally doesn't change our implementation of ChaCha20, just its interface so that it matches RFC8439. It otherwise behaves in the same way as previously. The new test vectors match those of the RFC.

theStack committed:
test: add ChaCha20 test triggering 32-bit block counter overflow

Verify that our ChaCha20 implementation using the 96/32 split interface
is compatible with >256 GiB outputs by triggering a 32-bit block counter
overflow and checking that the keystream matches one created with an
alternative implementation using a 64/64 split interface with the
corresponding input data. The test case data was generated with the
following Python script using the PyCryptodome library (version 3.15.0):

----------------------------------------------------------------------------------------------
from Crypto.Cipher import ChaCha20
key = bytes(list(range(32))); nonce = 0xdeadbeef12345678; pos = 2**32 - 1
c = ChaCha20.new(key=key, nonce=nonce.to_bytes(8, 'little'))
c.seek(pos * 64); stream = c.encrypt(bytes([0])*128)
print(f"Key: {key.hex()}\nNonce: {hex(nonce)}\nPos: {hex(pos)}\nStream: {stream.hex()}")
----------------------------------------------------------------------------------------------

theStack reviewed (commented):
Nice and indeed much simpler now, the actual diff is smaller than I expected. Will re-review tomorrow.

Meanwhile, I thought it would be nice to have a test case triggering the 32-bit block counter overflow to check for >256 GiB output compatibility and wrote one (test data is created with PyCryptodome, see commit message), maybe it's worthwhile to include it here: https://github.com/theStack/bitcoin/commit/61b41ad16e35306485a1db508d624d3ddab0ca0a (the next commit https://github.com/theStack/bitcoin/commit/ff289308bf7af8184cc0ecaafe6f5e1842b8f17b goes one step further and checks if the overflow really happened by inspecting the input state before/after test case execution, but not sure if it's worth it to pollute the ChaCha20 class with test-only methods).

theStack reviewed (approved):
Code-review ACK 0bf87476f55dceb106563156c7c8d6bfb8162e29

achow101 commented:
ACK 0bf87476f55dceb106563156c7c8d6bfb8162e29
